core-web/src/main/java/org/silverpeas/core/web/treemenu/process/TreeBuilder.java

```diff
@@ -194,45 +148,172 @@ private static MenuItem buildFirstLevel(TreeFilter filter, String userId, String
     MenuItem item = new MenuItem("root");
     ArrayList<MenuItem> children = new ArrayList<>();
     item.setChildren(children);
-    // the space
-    OrganizationController controller = OrganizationControllerProvider.getOrganisationController();
-    if (filter.acceptNodeType(NodeType.SPACE)) {
-      List<SpaceInstLight> rootSpaces =
-          controller.getRootSpacesContainingComponent(userId, "kmelia");
-      for (SpaceInstLight space : rootSpaces) {
-        if (space != null) {
-          MenuItem subElement =
-              new MenuItem(space.getName(language), space.getId(), 0,
-              NodeType.SPACE, false, null, null);
-          children.add(subElement);
-        }
-      }// the component
-    } else if (filter.acceptNodeType(NodeType.COMPONENT) && !filter.getComponents().isEmpty()) {
+
+    // the spaces
+    MenuItem subElement = getSpaceChildren(userId, language);
+    if (filter.acceptNodeType(NodeType.SPACE) && subElement != null) {
+      children.add(subElement);
+    }
+
+    // the components
+    if (filter.acceptNodeType(NodeType.COMPONENT) && !filter.getComponents().isEmpty()) {
       for (String componentName : filter.getComponents()) {
-        List<ComponentInstLight> componentList =
-            controller.getAvailComponentInstLights(userId, componentName);
-        // FIXME : externalize components specific in a properties
-        boolean isLeaf = getLeafValue(componentName);
-        for (ComponentInstLight compo : componentList) {
-          MenuItem subElement =
-              new MenuItem(Encode.forHtml(compo.getLabel(language)), compo.getId(), 0,
+        children.addAll(getNodesSubItem(userId, componentName, language));
+      }
+    }
+    return item;
+  }
+
+  /**
+   * Get space children
+   * @param userId the user identifier to display only the authorized item menu
+   * @param language the user language to display the label menu in the correct language
+   * @return a space level of the menu
+   */
+  private static MenuItem getSpaceChildren(String userId, String language)
+  {
+    List<SpaceInstLight> rootSpaces =
+        controller.getRootSpacesContainingComponent(userId, KMELIA);
+    if (TreeHandler.useOrder) {
+      rootSpaces.sort(Comparator.comparing(SpaceInstLight::getOrderNum));
+    }
+    for (SpaceInstLight space : rootSpaces) {
+      if (space != null) {
+        return new MenuItem(space.getName(language), space.getId(), 0,
+                NodeType.SPACE, false, null, null);
+      }
+    }
+    return null;
+  }
```

Comment:
The implementation of this method is wrong according to the code it replaces: the root spaces containing a Kmelia component instance should be browsed in order to create for each of them a MenuItem object and then all these created menu items should be returned to the caller

Reply:
I refactored as Sonar lint advices to reduce complexity code. It's probably possible to refactor more with more work.